_posts/2016-12-27-angular-2-advance-testing-with-custom-matchers.md

```diff
@@ -0,0 +1,650 @@
+{% endhighlight %}
+
+##### Special Helpers
+
+Earlier, I showed a code snapshot that had a *special helper* `activateMediaQuery( )`:
+
+![pic8](https://cloud.githubusercontent.com/assets/210413/21514513/8a9690ea-cc8b-11e6-98ac-2a66a42f023d.png)
+
+The Flex-Layout library has a responsive engine that supports change detection when a mediaQuery 
+activates (*aka* when the viewport size changes). The API uses selectors with dot-notation to 
+indicate which values should be used for which mediaQuery:
+
+{% highlight html %}
+{% raw %}
+<div fxLayout="row" fxLayout.md="column"></div>
+{% endraw %}
+{% endhighlight %}
+
+Testing these features presents several additional requirements. I need to be able to :
+
+* mock the window API `window.matchMedia(...)`
+* simulate a mediaQuery activation
+* trigger fixture.detectChange() after a simulated activation
+* hide all these details from indvidual tests (DRY)
+
+The Flex-Layout library actually publishes a MockMatchMedia class... and our earlier TestBed 
+configuration showed how I setup the DI:
+
+{% highlight js %}
+{% raw %}
+let fixture: ComponentFixture<any>;
+
+/**
+ * Special Helper
+ */
+let activateMediaQuery = (alias) => {
+      let matchMedia : MockMatchMedia = fixture.debugElement.injector.get(MatchMedia);
+      matchMedia.activate(alias);
+    };
+
+beforeEach(() => {
+
+  // Configure testbed to prepare services
+  TestBed.configureTestingModule({
+    imports: [FlexLayoutModule.forRoot()],
+    declarations: [TestLayoutComponent],
+    providers: [
+      BreakPointRegistry, BreakPointsProvider,
+      {provide: MatchMedia, useClass: MockMatchMedia}   //Special mock config
+    ]
+  })
+});
+{% endraw %}
+{% endhighlight %}
+
+Two (2) very interesting things are happening here: override DI providers and dynamic injection
+
+(1) Overriding DI Providers:
+
+Using
+
+{% highlight js %}
+{% raw %}
+{ provide: MatchMedia, useClass: MockMatchMedia }
+{% endraw %}
+{% endhighlight %}
+
+tells the TestBed DI system to provide an instance of the `MockMatchMedia` **whenever** any code 
+asks for an instance of the `MatchMedia` token to be injected via DI.
+
+(2) Dynamic Injection
+
+Our special helper `activateMediaQuery()` needs a dynamic injected instance of the MatchMedia token. 
+Using the `fixture` instance, I can dynamically get a **MockMatchMedia** instance from our fixtures 
+injector.
+
+Notice that all this complexity [and construction details] is encapsulated in our TestBed and special 
+helper... and the individual tests simply make the call:
+
+{% highlight js %}
+{% raw %}
+activatemediayQuery('md');
+{% endraw %}
+{% endhighlight %}
+
+That is very, very cool!
+
+### Custom Matchers
+
+I have only one more tool - in our testing toolkit - to discuss: Custom Jasmine Matchers.
+
+For those developers not familiar with the concepts of Jasmine *matchers*, I recommend the online 
+Jasmine documentation:
+
+<a href="https://jasmine.github.io/2.0/custom_matcher.html" target="_blank">
```

Comment:
Please add a `title` attribute and remove the `target` attribute. We let the user decide when she wants to open a link in a new tab or follow in same tab. More on that here: https://css-tricks.com/use-target_blank/